cvat/apps/engine/cloud_provider.py

```diff
@@ -176,51 +210,52 @@ def optimally_image_download(self, key: str, chunk_size: int = 65536) -> BytesIO
         Returns:
             BytesIO: Buffer with image
         """
-        image_parser=ImageFile.Parser()
+        image_parser = ImageFile.Parser()
 
         chunk = self.download_range_of_bytes(key, chunk_size - 1)
         image_parser.feed(chunk)
 
         if image_parser.image:
-            buff = BytesIO(chunk)
+            buff = NamedBytesIO(chunk)
+            buff.filename = key
         else:
             buff = self.download_fileobj(key)
             image_size_in_bytes = len(buff.getvalue())
             slogger.glob.warning(
                 f'The {chunk_size} bytes were not enough to parse "{key}" image. '
                 f'Image size was {image_size_in_bytes} bytes. Image resolution was {Image.open(buff).size}. '
                 f'Downloaded percent was {round(min(chunk_size, image_size_in_bytes) / image_size_in_bytes * 100)}')
-        buff.filename = key
+
         return buff
 
     def bulk_download_to_memory(
         self,
         files: List[str],
-        threads_number: int = min(get_cpu_number(), 4),
+        *,
+        threads_number: Optional[int] = None,
         _use_optimal_downloading: bool = True,
-    ) -> List[BytesIO]:
+    ) -> Iterator[BytesIO]:
         func = self.optimally_image_download if _use_optimal_downloading else self.download_fileobj
-        if threads_number > 1:
-            with ThreadPool(threads_number) as pool:
-                return pool.map(func, files)
-        else:
-            slogger.glob.warning('Download files to memory in series in one thread.')
-            return [func(f) for f in files]
+        threads_number = normalize_threads_number(threads_number, len(files))
+
+        with ThreadPoolExecutor(max_workers=threads_number) as executor:
+            yield from executor.map(func, files)
 
     def bulk_download_to_dir(
         self,
         files: List[str],
         upload_dir: str,
-        threads_number: int = min(get_cpu_number(), 4),
-    ):
-        args = zip(files, [os.path.join(upload_dir, f) for f in files])
-        if threads_number > 1:
-            with ThreadPool(threads_number) as pool:
-                return pool.map(lambda x: self.download_file(*x), args)
-        else:
-            slogger.glob.warning(f'Download files to {upload_dir} directory in series in one thread.')
-            for f, path in args:
-                self.download_file(f, path)
+        *,
+        threads_number: Optional[int] = None,
+    ) -> None:
+        threads_number = normalize_threads_number(threads_number, len(files))
+
+        with ThreadPoolExecutor(max_workers=threads_number) as executor:
+            futures = [executor.submit(self.download_file, f, os.path.join(upload_dir, f)) for f in files]
+            done, _ = wait(futures, return_when=FIRST_EXCEPTION)
+            for future in done:
+                if ex := future.exception():
+                    raise ex
```

Comment:
I feel that this implementation can store all the thread results in memory

Given a simple script like this:
```python
from concurrent.futures import ThreadPoolExecutor
import sys
from time import sleep


class ObjectWithAllocationReporting:
    def __init__(self, i) -> None:
        self.id = i
        print("created", self.id)

    def __del__(self) -> None:
        print("deleted", self.id)

def main() -> int:
    def _foo_with_result(i: int) -> ObjectWithAllocationReporting:
        return ObjectWithAllocationReporting(i)

    def _foo_without_result(i: int) -> None:
        ObjectWithAllocationReporting(i)
        # can also be a wrapper around a function with a result
        # can also add del here to be sure

    for callback in [
        _foo_with_result,
        _foo_without_result,
    ]:
        print(callback)
        with ThreadPoolExecutor(2) as pool:
            results = pool.map(callback, range(6))
            sleep(5)
            print("slept")
            list(results)

    return 0

if __name__ == "__main__":
    sys.exit(main())
```

Output:
```
<function main.<locals>._foo_with_result at 0x7fc98bd54e50>
created 0
created 1
created 2
created 3
created 4
created 5
slept
deleted 5
deleted 4
deleted 3
deleted 2
deleted 1
deleted 0
<function main.<locals>._foo_without_result at 0x7fc98bd54ee0>
created 0
deleted 0
created 1
deleted 1
created 2
deleted 2
created 3
created 4
deleted 3
deleted 4
created 5
deleted 5
slept
```

Probably, it's not what we want for big datasets, as there is going to be some time between the tasks are started and the results are collected.